c7n/resources/glue.py

```diff
@@ -184,6 +184,54 @@ class resource_type(TypeInfo):
     augment = universal_augment
 
 
+@GlueCrawler.filter_registry.register('security-config')
+class GlueCrawlerSecurityConfigFilter(ValueFilter):
+    """Filters glue crawlers with security configurations
+
+    :example:
+
+    .. code-block:: yaml
+
+            policies:
+              - name: need-kms-cloudwatch
+                resource: glue-crawler
+                filters:
+                  - type: security-config
+                    key: EncryptionConfiguration.CloudWatchEncryption.CloudWatchEncryptionMode
+                    op: ne
+                    value: SSE-KMS
+    """
+
+    permissions = ('glue:GetSecurityConfiguration',)
+    sec_conf_attribute = 'c7n:SecurityConfiguration'
+    security_config_key = 'CrawlerSecurityConfiguration'
+    schema = type_schema('security-config', rinherit=ValueFilter.schema)
+
+    def process(self, resources, event=None):
+        self.log.debug("fetching security configuration for %d glue crawlers" % len(resources))
+        resources = self.get_security_configuration(resources)
+        return super(ValueFilter, self).process(resources, event)
+
+    def __call__(self, r):
+        if self.sec_conf_attribute not in r:
+            return False
+        return self.match(r[self.sec_conf_attribute])
+
+    def get_security_configuration(self, resources):
+        client = local_session(self.manager.session_factory).client('glue')
+        for r in resources:
+            if self.security_config_key not in r:
```

Comment:
this conditional looks suspect,  what is security_config_key and why would it be in a crawler without querying it (sec_conf_attribute)?

Reply:
Thanks for the input. I don't really understand the subtleties of how you're supposed to append 'c7n:*' data on your resources in custodian. With that in mind I'll explain my reasoning on the two **if** statements. I also renamed security_config_key to boto_security_config_key to hopefully make the code less confusing. 

boto_security_config_key is the json key name that is returned by the get_crawlers() api call. boto_security_config_key points to the name of the security configuration. Unfortunately, it is named differently for different glue resources. 'CrawlerSecurityConfiguration' is used for glue crawlers and 'SecurityConfiguration' is used for glue dev endpoints. The interesting thing is security configurations are optional and when they are absent the 'CrawlerSecurityConfiguration' is left out the returned json (as oppose to 'CrawlerSecurityConfiguration' having an empty value).  I put this as a variable to eventually use it as a base class and extend with the appropriate value.


I put the check:

**if self.sec_conf_attribute not in r:**

for the possible of a filter a policy like:

    - name: get-glue-crawler-filter
      resource: glue-crawler
      filters:
          - type: security-config
            key: EncryptionConfiguration.CloudWatchEncryption.CloudWatchEncryptionMode
            op: eq
            value: SSE-KMS
          - type: security-config
            key: EncryptionConfiguration.JobBookmarksEncryption.JobBookmarksEncryptionMode
            op: eq
            value: CSE-KMS

My thought process was to save extra api calls to get_security_configuration() on the 2nd filter. All the security information would already be cached from the first filter. 

I did not understand your second sentence:
" if its is present on the resource then processing of the resource will fail on key error because sec_conf_attribute won't be present since its not being populated."
Could you explain it in a slightly different way? Thank you

Reply:
the way i read this code, you had two keys, A and B in a dict, you had if A exists skip setting B, but then went to process B unconditionally even though it wasn't set.

Reply:
@kapilt I think I addressed your above comment with the last commit.

Reply:
renaming the variables for clarity doesn't address the logic flaw i'm commenting on. you'd have to filter the resources that  don't have attribute B so things don't have exceptions later when processing resources based on attribute B